Source/Foundation/bsfCore/BsCoreApplication.cpp

```diff
@@ -231,108 +233,129 @@ namespace bs
 				mLastFrameTime = currentTime;
 			}
 
-			gProfilerCPU().beginThread("Sim");
+			runMainLoopFrame();
+		}
 
-			Platform::_update();
-			DeferredCallManager::instance()._update();
-			gTime()._update();
-			gInput()._update();
-			// RenderWindowManager::update needs to happen after Input::update and before Input::_triggerCallbacks,
-			// so that all input is properly captured in case there is a focus change, and so that
-			// focus change is registered before input events are sent out (mouse press can result in code
-			// checking if a window is in focus, so it has to be up to date)
-			RenderWindowManager::instance()._update();
-			gInput()._triggerCallbacks();
-			gDebug()._triggerCallbacks();
+		// Wait until last core frame is finished before exiting
+		waitUntilFrameFinished();
+	}
 
-			preUpdate();
+	void CoreApplication::beginMainLoop()
+	{
+		mRunMainLoop = true;
+	}
 
-			// Trigger fixed updates if required
-			{
-				UINT64 step;
-				const UINT32 numIterations = gTime()._getFixedUpdateStep(step);
+	void CoreApplication::endMainLoop()
+	{
+		// One more frame needs to be rendered here to not crash if the outer main loop has done stuff since the last frame.
+		runMainLoopFrame();
```

Comment:
This bit is still bothering me. What exactly happens when you remove it? Does the crash happen with normal bsf code or only your app (e.g. does it happen in examples)? If it's only your code then I'd be happiest we removed it here.

Reply:
I removed that call. I'm sure though that if anyone uses the single frame functions like I do (with releasing objects after the last frame), they are going to need to have the extra frame added in their code.